### PR TITLE
Implement hint for the next block

### LINKED ISSUE #12: Hint about the next block
## User story

```
As a player,
I want to see the next block that is going to appear after the current one,
So that I can plan my moves better.
```

## Acceptance criteria

1. The user interface shows the next block to appear on the playfield.
2. When the current block settles at the bottom of the playfield:
   * the hinted block becomes the new current one in the next round,
   * a new block appears as hinted.

### PR DESCRIPTION
**Pull request by kweut**
Hint block is printed inside a frame above playfield.

Closes #12

### CODE REVIEW

File: src/main/java/com/epam/prejap/teatrees/game/Playfield.java [hunk 1]
```diff
@@ -109,4 +111,18 @@ private void forEachBrick(BrickAction action) {
...
+    Block getBlock() {
+        return block;
+    }
+
+    /**
+     * Method created only for testing.
+     */
+    Block getHintBlock() {
+        return hintBlock;
+    }
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/game/Playfield.java:126]: If you really need access to `private` fields, then rather reflection should be your thing. Here is some useful material for you: http://tutorials.jenkov.com/java-reflection/index.html

Yet, probably you should have created a test for `nextBlock()`, hijack output and compare it with expected. There may be a problem with randomness.

File: src/main/java/com/epam/prejap/teatrees/game/Printer.java [hunk 2]
```diff
@@ -1,5 +1,7 @@
 package com.epam.prejap.teatrees.game;
 
+import com.epam.prejap.teatrees.block.Block;
+
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/game/Printer.java:4]: unnecessary empty line

File: src/main/java/com/epam/prejap/teatrees/game/Printer.java [hunk 3]
```diff
@@ -10,17 +12,43 @@ public Printer(PrintStream out) {
...
             startRow();
             for (byte aByte : bytes) {
                 print(aByte);
             }
             endRow();
         }
-        border(grid[0].length);
+    }
+
+    void spaces(int times) {
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/game/Printer.java:50]: Will this be used somewhere outside? if not, then maybe `private`?

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    public void shouldHintBlockBecomeCurrentlyPlayedBlock() {
+        // given
+        Playfield playfield = new Playfield(10, 10, new BlockFeed(), new FakePrinter());
+        Block hintBlock = playfield.getHintBlock();
+
+        // when
+        playfield.nextBlock();
+
+        // then
+        assertTrue(hintBlock == playfield.getBlock());
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java:20]: `assetEquals`

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    }
+}
+
+class FakePrinter extends Printer {
+    FakePrinter() {
+        super(System.out);
+    }
+
+    @Override
+    void draw(byte[][] grid, Block hintBlock) {}
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java:30]: Are you using this?

File: src/main/java/com/epam/prejap/teatrees/game/Playfield.java (hunk 1 above)
Comment (kweut) [reply src/main/java/com/epam/prejap/teatrees/game/Playfield.java]: Thanks for the link, it was very useful :)
These methods are removed now.

File: src/main/java/com/epam/prejap/teatrees/game/Printer.java (hunk 2 above)
Comment (kweut) [reply src/main/java/com/epam/prejap/teatrees/game/Printer.java]: Removed.

File: src/main/java/com/epam/prejap/teatrees/game/Printer.java (hunk 3 above)
Comment (kweut) [reply src/main/java/com/epam/prejap/teatrees/game/Printer.java]: I changed other methods in this class to private as well.

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java (hunk 4 above)
Comment (kweut) [reply src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java]: Changed.

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java (hunk 5 above)
Comment (kweut) [reply src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java:49]: Yes, but not directly.
It goes like this: Playfield::nextBlock calls Playfield::show and the latter calls Printer::draw which prints hint and playfield.
The goal of this test is to check block swaping mechanism so there is no point in printing anything. Thus I pass to Playfield's contructor an instance of a FakePrinter which doesn't print anything.

Comment (marcin-ogorzalek-EPAM) [reply src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java:305]: Ok, then this is one of the rare occasions when tests should have javadocs. Create javadoc for this fake printer explaining what we have here.

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java [hunk 6]
```diff
@@ -262,4 +265,42 @@ public void rotateBlockImpossible() {
...
+        Field[] fields = playfieldClass.getFields();
+        try {
+            for (Field field : fields) {
+                if (field.getName().equals(block)) {
+                    field.setAccessible(true);
+                    hintBlock = field.get(playfield);
+                }
+            }
+        } catch (IllegalAccessException e) {
+            System.err.println("Illegal access.");
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java:293]: This should be declared with `throws`. In tests, this is an acceptable solution, as we wish for those to stop executing ASAP a point the problem (if there is any).

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java (hunk 5 above)
Comment (kweut) [reply src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java:310]: Done.

File: src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java (hunk 6 above)
Comment (kweut) [reply src/test/java/com/epam/prejap/teatrees/game/PlayfieldTest.java]: Done.

**Pull request merged**